snapshots/devmapper/pool_device.go

```diff
@@ -155,7 +154,7 @@ func (p *PoolDevice) transition(ctx context.Context, deviceName string, tryingSt
 		} else {
 			deviceInfo.Error = err.Error()
 		}
-		return nil
+		return err
```

Comment:
I think we should not. In case of error this will rollback write transaction to metadata storage, which we don't want to.

Reply:
Looking at the code in PoolDevice#transition after the metadata.UpdateDevice call:
```
	if uerr != nil {
		result = multierror.Append(result, uerr)
```
It seems uerr would always be nil.
The above check is ineffective.

Reply:
`UpdateDevice` itself can return errors, right? https://github.com/containerd/containerd/blob/3a1c6a260a0a3f518593da4a6091dd447ab10260/snapshots/devmapper/metadata.go#L232

Reply:
The fn below:
```
		if err := fn(device); err != nil {
```
is the lambda I touched which returns nil.
There is no other change after the fn() call that would result in non-nil error being returned from UpdateDevice call.
To verify this, inserting a panic in the if block would do:
```
	if uerr != nil {
		result = multierror.Append(result, uerr)
```